feat: support rustflags per profile

zhamlin opened the pull request:
Fix for issue https://github.com/rust-lang/cargo/issues/7878

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented on src/cargo/core/profiles.rs:
```diff
@@ -666,6 +669,7 @@ pub struct Profile {
     pub incremental: bool,
     pub panic: PanicStrategy,
     pub strip: Strip,
+    pub rustflags: Option<InternedString>,
```
I think that this would be stored as `Vec<InternedString>` to avoid the join-with-space followed by a split-on-space.

zhamlin replied:
I was going to go with that approach but got an error about not being able to implement `Copy` for `Profile`.
I can change it to something else, but not sure what else would be preferred.

alexcrichton replied:
I'm not sure how to best fix that myself, but we can't join/split on spaces because that breaks arguments with spaces in them (they would become multiple arguments instead of one)

alexcrichton reviewed (commented):
Thanks for the PR for this! I think there's also some documentation in `src/doc` which may need updating as well.

Given that this is somewhat of a major feature as well I think I would personally prefer to err on the side of placing this behind a feature gate to start with. Others on the team may feel differently though!

ehuss commented:
I think it would be good to also talk about how this works with rustdocflags. If there are flags that are required for correctness, then `cargo test --doc` would fail.  I'm not suggesting that this should add rustdocflags, though. Just to consider what the impact will be.

I'm a little nervous adding this, as it is a very low-level thing which we haven't exposed in `Cargo.toml` before. One of my biggest concerns is that it can cause backward-compatibility hazards. For example, when Cargo adds new features or changes the way it invokes rustc, that can interfere with pre-existing flags.  In a sense, this seems to be an end-run around Cargo not exposing the workflows that users want.

I'm not concerned enough to block it, as I can see its utility. However, I think we should tread very carefully.

alexcrichton commented:
I share many of @ehuss's concerns as well, which is primarily why I'd like to see this feature-gated to start off with.

joshtriplett commented:
I agree that this needs a feature-gate initially.

alexcrichton commented on src/doc/src/reference/unstable.md:
```diff
@@ -842,6 +843,23 @@ You can also configure this option with the two absolute boolean values
+# ...
+
+[profile.release]
+rustflags = [ "-C" "..." ]
```
I think there needs to be a `,` here?

zhamlin replied:
Added

alexcrichton commented on tests/testsuite/profile_config.rs:
```diff
@@ -173,6 +173,38 @@ found package specs: bar, bar:0.5.0",
+
+    p.cargo("build -v")
+        .masquerade_as_nightly_cargo()
+        .env_remove("CARGO_INCREMENTAL")
```
Was this line necessary to get this to pass? If not then this should be removed, if so then this should be built in to other parts of cargo's test harness. IIRC there's already a function that does a bit of sanitization for the environment of child proceses.

zhamlin replied:
Assuming you were referring to `env_remove("CARGO_INCREMENTAL")`, it was not needed.

alexcrichton reviewed (commented):
Can you add a test that this is unstable?

Additionally I think that support also needs to be added for `CARGO_PROFILE_RELEASE_RUSTFLAGS=...` by configuring profiles via env vars.

bors commented:
:umbrella: The latest upstream changes (presumably #10088) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

zhamlin commented:
Thanks for the review.

Fixed the issues you mentioned and added a test to verify passing rustflags via `CARGO_PROFILE_RELEASE_RUSTFLAGS`, it appears to be working without any additional code changes.

alexcrichton commented:
Looks good to me, thanks! I'd like to double-check though that others are ok merging this as well (note that this implementation is gated behind a cargo feature)

@rfcbot fcp merge

ehuss commented on src/cargo/core/profiles.rs:
```diff
@@ -666,6 +669,8 @@ pub struct Profile {
     pub panic: PanicStrategy,
     pub strip: Strip,
+    // Note that `rustflags` is used for the cargo-feature `profile_rustflags`
+    pub rustflags: Vec<InternedString>,
```
Can this be excluded until it is stable?  I think adding `#[serde(skip_serializing_if = "Vec::is_empty")]` should be sufficient (with a comment to make it clear it should be removed when stabilized).

ehuss commented on src/doc/src/reference/unstable.md:
````diff
@@ -820,6 +821,23 @@ The following is a description of the JSON structure:
 ```
 
+### Profile `rustflags` option
+* Tracking Issue: [rust-lang/cargo#7878](https://github.com/rust-lang/cargo/issues/7878)
````
I opened a new tracking issue to more clearly track the progress for this.

```suggestion
* Original Issue: [rust-lang/cargo#7878](https://github.com/rust-lang/cargo/issues/7878)
* Tracking Issue: [rust-lang/cargo#10271](https://github.com/rust-lang/cargo/issues/10271)
```

ehuss commented:
I was a little concerned about the large number of new `clone()` calls, but in some rough benchmarking it doesn't seem to have much of an impact.

ehuss commented:
Thanks!

@bors r+

bors commented:
:hourglass: Testing commit 68fd67319dc766ff5b173989da2b72430857ee5d with merge 0655dd25a0242c56eb51faecc86dd67a5abfb8dd...
<!-- homu: {"type":"BuildStarted","head_sha":"68fd67319dc766ff5b173989da2b72430857ee5d","merge_sha":"0655dd25a0242c56eb51faecc86dd67a5abfb8dd"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/4765001654?check_suite_focus=true)
Approved by: ehuss
Pushing 0655dd25a0242c56eb51faecc86dd67a5abfb8dd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/4765001654?check_suite_focus=true"},"merge_sha":"0655dd25a0242c56eb51faecc86dd67a5abfb8dd"} -->
